Feature/milc mom

maddyscientist opened the pull request:
This pull request is aimed at improving the performance of MILC RHMC:
* Adds new QUDA interface function `momResidentQuda` for setting or returning the resident momentum field in QUDA
* Adds new MILC-QUDA interface functions `qudaMomLoad` and `qudaMomSave` which downloads and sets the resident momentum field, or uploads and returns the resident momentum field, respectively.  Prior behaviour is preserved with this PR, so older versions of MILC are not broken.  A new version of MILC will be required to enable the momentum residency.
* When `qudaMomLoad` is called, the gauge force, fermion force, action evaluation and momentum exponentiation will all utilize the resident momentum field.  Calling `qudaMomSave` will invalidate the resident field.
* Some cleanup and fixes to the the momentum action evaluation to make the code cleaner.
* Added host unit test variant of the momentum action and utilize this in `gauge_force_test`.
* Cleaned up `computeGaugeForceQuda` to now support QDP-ordered momentum fields removing the prior hacks.
* Clean up of `gauge_force_test` to make the code cleaner and remove unnecessarily reimplementations of data reordering and field extension.

maddyscientist committed:
Ensure that MILC site params are printed with printQudaGaugeParam

maddyscientist committed:
Add support for momActionQuda to use MILC site structures.  Added new QUDA function which will either set the resident momentum or return it

maddyscientist committed:
Add site struct support to qudaMomAction.  Added two new prototype functions, qudaMomLoad and qudaMomSave for downloading and uploading the resident momentum field from MILC

maddyscientist committed:
Add updated milc_interface.cpp, which was missed in last commit

maddyscientist committed:
Cleanup computeMomAction kernel

maddyscientist committed:
Momentum residency support for MILC.  When qudaMomLoad is called the momentum field is downloaded to the GPU and made resident.  Subsequent kernels that use the momentum (action, gauge force, fermion force, gauge evolution) will use the resident momentum, avoiding host-device transfers

maddyscientist committed:
Add support for recon-10 cpuGaugeFields with QDP ordering

maddyscientist committed:
Fix hacks in computeGaugeForceQuda to support QDP-orderfor momentum fields and cleanup to use QUDA internal copy and exchange routines.  Gauge force unit test now includes a test for momActionQuda.

maddyscientist committed:
Apply clang-format

mathiaswagner commented on tests/gauge_force_test.cpp:
```diff
@@ -366,63 +265,100 @@ void gauge_force_test(void)
     }
   }
 
+  printfQuda("%d\n", __LINE__);
```
should the print statements be here ?

mathiaswagner reviewed (changes requested):
Just wondering about some printfs in the test. Verified with MILC RHMC. Looks good to me.

maddyscientist committed:
Remove debug printfs and add correctness check for momentum action

maddyscientist committed:
Add initial gtest support to gauge_force_test and add to ctest list with gauge-force is enabled.  Fixed but in multi-process reference mom-action evaluation

maddyscientist commented:
Thanks for the 👀  @mathiaswagner.  I have 
* removed the debug printf statements
* fixed a bug with reference mom action function
* added initial gtest/ctest support so that the gauge force will now be tested with `ctest` invocation.  E.g.,
```
ctest -N
Test project /scratch/kate/quda-develop
  Test #1: blas_test_parity_staggered
  Test #2: blas_test_full_staggered
  Test #3: dslash_improved_staggered-policytune
  Test #4: dslash_naive_staggered-policytune
  Test #5: dslash_improved_staggered_build-policytune
  Test #6: gauge_force
```
This should be good to merge now.

weinbe2 reviewed (approved):
Looks good to me. I'm happy to see a gauge force `ctest` was added.

Visual review only. Didn't see a need to drag it through the mud because it doesn't touch solvers, blas, dslash, etc.
